From ae5fd4a77b03b1a3719f71857acb7205931813f4 Mon Sep 17 00:00:00 2001
From: Suman Anna <s-anna@ti.com>
Date: Fri, 20 Feb 2015 15:35:41 -0600
Subject: [PATCH 09/13] remoteproc: add macros to use with rproc_da_to_va()
 flags

The rproc_da_to_va() API takes in a flags argument to pass
it onto the individual remoteproc implementations to aid them
in performing proper address translations on internal memory
regions. As this can vary from one implementation to the other,
define some semantics for this flags argument.

A new flags type is encoded into the upper 16 bits along side
the actual value in the lower 16-bits for the flags argument,
to allow different individual implementations to have better
flexibility in interpreting the flags as per their needs.

The current remoteproc ELF loader code is also updated to
use this semantics to pass the different types of flags used
in different invocations of rproc_da_to_va().

Signed-off-by: Suman Anna <s-anna@ti.com>
---
 drivers/remoteproc/remoteproc_core.c       |    6 ++++--
 drivers/remoteproc/remoteproc_elf_loader.c |    5 +++--
 drivers/remoteproc/remoteproc_internal.h   |   14 ++++++++++++++
 3 files changed, 21 insertions(+), 4 deletions(-)

diff --git a/drivers/remoteproc/remoteproc_core.c b/drivers/remoteproc/remoteproc_core.c
index 9e8ed61..0a839d0 100644
--- a/drivers/remoteproc/remoteproc_core.c
+++ b/drivers/remoteproc/remoteproc_core.c
@@ -162,7 +162,9 @@ static void rproc_disable_iommu(struct rproc *rproc)
  * so we can access the referenced memory. This function also allows to perform
  * translations on the internal remoteproc memory regions through a platform
  * implementation specific da_to_va ops, if present. The @flags field is passed
- * onto these ops to aid the translation within the ops implementation.
+ * onto these ops to aid the translation within the ops implementation. The
+ * @flags field is to be passed as a combination of the RPROC_FLAGS_xxx type
+ * and the pertinent flags value for that type.
  *
  * The function returns a valid kernel address on success or NULL on failure.
  *
@@ -581,7 +583,7 @@ static int rproc_handle_trace(struct rproc *rproc, struct fw_rsc_trace *rsc,
 	}
 
 	/* what's the kernel address of this resource ? */
-	ptr = rproc_da_to_va(rproc, rsc->da, rsc->len, 0);
+	ptr = rproc_da_to_va(rproc, rsc->da, rsc->len, RPROC_FLAGS_NONE);
 	if (!ptr) {
 		dev_err(dev, "erroneous trace resource entry\n");
 		return -EINVAL;
diff --git a/drivers/remoteproc/remoteproc_elf_loader.c b/drivers/remoteproc/remoteproc_elf_loader.c
index c7d2cf3..871dd11 100644
--- a/drivers/remoteproc/remoteproc_elf_loader.c
+++ b/drivers/remoteproc/remoteproc_elf_loader.c
@@ -183,7 +183,8 @@ rproc_elf_load_segments(struct rproc *rproc, const struct firmware *fw)
 		}
 
 		/* grab the kernel address for this device address */
-		ptr = rproc_da_to_va(rproc, da, memsz, phdr->p_flags);
+		ptr = rproc_da_to_va(rproc, da, memsz,
+				     RPROC_FLAGS_ELF_PHDR | phdr->p_flags);
 		if (!ptr) {
 			dev_err(dev, "bad phdr da 0x%x mem 0x%x\n", da, memsz);
 			ret = -EINVAL;
@@ -326,7 +327,7 @@ rproc_elf_find_loaded_rsc_table(struct rproc *rproc, const struct firmware *fw)
 		return NULL;
 
 	return rproc_da_to_va(rproc, shdr->sh_addr, shdr->sh_size,
-				shdr->sh_flags);
+			      RPROC_FLAGS_ELF_SHDR | shdr->sh_flags);
 }
 
 /**
diff --git a/drivers/remoteproc/remoteproc_internal.h b/drivers/remoteproc/remoteproc_internal.h
index e0f09bc..4e4bb83 100644
--- a/drivers/remoteproc/remoteproc_internal.h
+++ b/drivers/remoteproc/remoteproc_internal.h
@@ -22,6 +22,20 @@
 
 #include <linux/irqreturn.h>
 #include <linux/firmware.h>
+#include <linux/bitops.h>
+
+/*
+ * Macros to use with flags field in rproc_da_to_va API. Use
+ * the upper 16 bits to dictate the flags type and the lower
+ * 16 bits to pass on the value of the flags pertinent to that
+ * type.
+ *
+ * Add any new flags type at a new bit-field position
+ */
+#define RPROC_FLAGS_SHIFT	16
+#define RPROC_FLAGS_NONE	0
+#define RPROC_FLAGS_ELF_PHDR	BIT(0 + RPROC_FLAGS_SHIFT)
+#define RPROC_FLAGS_ELF_SHDR	BIT(1 + RPROC_FLAGS_SHIFT)
 
 struct rproc;
 
-- 
1.7.9.5

